### PR TITLE
Reduce time to run tests by using a persistent source schema

### PR DESCRIPTION
**Pull request by plypaul**
This PR implements the issue https://github.com/dbt-labs/metricflow/issues/482 and allows the user to run `pytest --use-persistent-source-schema ...`.

### CODE REVIEW

**Review by tlento — COMMENTED** [review]
@plypaul the first Databricks run failed due to concurrent table creation. I have a suspicion this is because we run both the cluster and SQL warehouse modes, but I'm not 100% sure on this.

If that is the case, these tests will likely fail whenever someone changes the data and creates a new schema, but if a re-run fixes it I think that's ok.

File: metricflow/test/fixtures/setup_fixtures.py [hunk 1]
```diff
@@ -67,7 +83,23 @@ def mf_test_session_state(request: FixtureRequest) -> MetricFlowTestSessionState
...
-    mf_source_schema = mf_system_schema
+
+    use_persistent_source_schema = bool(request.config.getoption(USE_PERSISTENT_SOURCE_SCHEMA_CLI_FLAG, default=False))
+    if use_persistent_source_schema:
+        source_table_snapshots_hash = SqlTableSnapshotHash.create_from_hashes(
+            tuple(table_snapshot.snapshot_hash for table_snapshot in source_table_snapshot_repository.table_snapshots)
+        )
+        # Since the source schema is used in the SQL output snapshots (but replaced with *'s), make the name of the
+        # schema to be the same length so that the SQL output snapshots don't change.
+        mf_source_schema = f"mf_test_src_{source_table_snapshots_hash.str_value}"[: len(mf_system_schema)]
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/test/fixtures/setup_fixtures.py]: Added a comment and also added some safety checks with the length.

File: .github/workflows/cd-sql-engine-tests.yaml [hunk 2]
```diff
@@ -36,7 +36,7 @@ jobs:
         run: cd metricflow && poetry install
 
       - name: Run MetricFlow unit tests with Snowflake configs
-        run: poetry run pytest metricflow/test/ --ignore=metricflow/test/model/dbt_cloud_parsing
+        run: poetry run pytest -v --use-persistent-source-schema metricflow/test/ --ignore=metricflow/test/model/dbt_cloud_parsing
```
[the remark this answers is not in the document]
Comment (plypaul) [reply .github/workflows/cd-sql-engine-tests.yaml:39]: The verbose mode shows a better representation of test progress in the logs. Updated the commit message.

File: .changes/unreleased/Features-20230428-172214.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+kind: Features
```
[the remark this answers is not in the document]
Comment (plypaul) [reply .changes/unreleased/Features-20230428-172214.yaml]: Updated.

**Comment by plypaul** [comment]
> @plypaul the first Databricks run failed due to concurrent table creation. I have a suspicion this is because we run both the cluster and SQL warehouse modes, but I'm not 100% sure on this.

> If that is the case, these tests will likely fail whenever someone changes the data and creates a new schema, but if a re-run fixes it I think that's ok.

Yeah, a re-run would fix it. Without locking / concurrency controls, you would get that kind of behavior (or incorrectly loaded tables) if there are two concurrent test runs getting launched with an updated dataset. It's hard to resolve without adding a bit of complexity, and given that it should be rare, I didn't think it was worthwhile to add it in. Flakey CI is terrible though, so I'm going to keep an eye on it and see if we need to do something about it.

**Comment by plypaul** [comment]
@tlento Can't find the log now, but I think the Databricks failure that you saw was caused by a test that created temporary tables in the source schema. It should have created tables in the ephemeral system schema. Updated that test.

**Pull request merged**